File: manager/src/keymap.js
// KeyMap class for each key map item.
class KeyMap {
    constructor() {
        this.keyCode = null; // actual trigger key code.
        this.displayKey = null; // display key just for display in configuration page.
        this.targetApp = ""; // target process name for key mapping.
        this.map = {
            // action keys mapped to trigger key
            keyCode: null,  // KeyboardEvent.keyCode. Virtual key code
            displayKey: null,  // KeyboardEvent.key (could be modified)
            // modifier keys for key combination
            modifierKeys: {
                ctrl: false,
                shift: false,
                alt: false,
                win: false,
            },
            // initialize map
            init() {
                this.keyCode = null;
                this.displayKey = null;
                this.modifierKeys.ctrl = false;
                this.modifierKeys.shift = false;
                this.modifierKeys.alt = false;
                this.modifierKeys.win = false;
            },
        };
    }
}

module.exports = KeyMap;

File: engine/src/KeyMap.cpp
#include "KeyMap.h"


map<pair<int, string>, KeyMap> keyMaps;

// keyCode 0 means key mapping with the keyCode does not exist.
KeyMap::KeyMap() : keyCode(0), map{0, "", {false, false, false, false}} {}

KeyMap::KeyMap(const nlohmann::json& jsonData) {
    keyCode = jsonData["keyCode"].get<int>();
    displayKey = jsonData["displayKey"].get<string>();
    targetApp = jsonData["targetApp"].get<string>();

    const nlohmann::json& mapData = jsonData["map"];
    map.keyCode = mapData["keyCode"].get<int>();
    map.displayKey = mapData["displayKey"].get<string>();

    const nlohmann::json& modifierKeysData = mapData["modifierKeys"];
    map.modifierKeys.ctrl = modifierKeysData.value("ctrl", false);
    map.modifierKeys.shift = modifierKeysData.value("shift", false);
    map.modifierKeys.alt = modifierKeysData.value("alt", false);
    map.modifierKeys.win = modifierKeysData.value("win", false);
}